### ISSUE TITLE
container cuda driver version conflict

### ISSUE THREAD

**Issue by better629** [alibaba/x-deeplearning#247]
The image `ubuntu-gpu-mxnet1.3` is built on cuda9.0，but the host installed the cuda runtime driver 418.* and cuda10.1.
Will it be a comflict because of the error that `CUDA driver version is insufficient for CUDA runtime version` in docker.

**Comment by better629** [comment]
This is a error with docker `DOCKER_PATH='registry.cn-hangzhou.aliyuncs.com/xdl/xdl:ubuntu-gpu-mxnet1.3'`

My Host machine with multi-gpus which run ./deviceQuery and nvidia-smi is fine.

the container was started with `docker run -it --net=host --volume $HOME:$HOME -w $HOME `curl -s http://localhost:3476/docker/cli` --name tdm-mock "$DOCKER_PATH"` in `wiki/深度树匹配模型(TDM)`. I also tried `docker run --runtime=nvidia ***` and `nvidia-docker run ****`, the error is the same.


with `tdm_ub_att_ubuntu`example, run `python train.py --run_mode=local --config=config.train.json`, output is below:  

```
concat
call FullyConnected3D
in FullyConnected3D use prelu
call FullyConnected3D
in FullyConnected3D use prelu
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
[WARNING]: no gradient found for var:bn_1_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_1_d_moving_var under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_2_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_2_d_moving_var under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_3_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_3_d_moving_var under scope: , maybe not used?
>>>>>>>>>>>> 0 >>>>>>>>>>>
Cuda Stream Create Error Error Code: Aborted (core dumped)
```

So I change the `gpu GPU` in train.py to `cpu CPU`, rerun. The output is below:
```
call FullyConnected3D
in FullyConnected3D use prelu
call FullyConnected3D
in FullyConnected3D use prelu
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
[WARNING]: no gradient found for var:bn_1_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_1_d_moving_var under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_2_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_2_d_moving_var under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_3_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_3_d_moving_var under scope: , maybe not used?
>>>>>>>>>>>> 0 >>>>>>>>>>>
save checkpoint at global_step[2], ckpt version[ckpt-...................2]
hdfsOpenFile(hdfs://{ip}:{host}/user/root/tdm/train/checkpoint/checkpoints): FileSystem#open((Lorg/apache/hadoop/fs/Path;I)Lorg/apache/hadoop/fs/FSDataInputStream;) error:
java.io.FileNotFoundException: File does not exist: /user/root/tdm/train/checkpoint/checkpoints
    at org.apache.hadoop.hdfs.server.namenode.INodeFile.valueOf(INodeFile.java:71)
    at org.apache.hadoop.hdfs.server.namenode.INodeFile.valueOf(INodeFile.java:61)
    at org.apache.hadoop.hdfs.server.namenode.FSDirStatAndListingOp.getBlockLocations(FSDirStatAndListingOp.java:152)
    at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.getBlockLocations(FSNamesystem.java:1750)
    at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.getBlockLocations(NameNodeRpcServer.java:691)
    at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.getBlockLocations(ClientNamenodeProtocolServerSideTranslatorPB.java:377)
    at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
    at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:447)
    at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:989)
    at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:845)
    at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:788)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:422)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1807)
    at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2455)

    at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
    at org.apache.hadoop.ipc.RemoteException.instantiateException(RemoteException.java:121)
    at org.apache.hadoop.ipc.RemoteException.unwrapRemoteException(RemoteException.java:88)
    at org.apache.hadoop.hdfs.DFSClient.callGetBlockLocations(DFSClient.java:828)
    at org.apache.hadoop.hdfs.DFSClient.getLocatedBlocks(DFSClient.java:815)
    at org.apache.hadoop.hdfs.DFSClient.getLocatedBlocks(DFSClient.java:804)
    at org.apache.hadoop.hdfs.DFSInputStream.fetchLocatedBlocksAndGetLastBlockLength(DFSInputStream.java:319)
    at org.apache.hadoop.hdfs.DFSInputStream.openInfo(DFSInputStream.java:280)
    at org.apache.hadoop.hdfs.DFSInputStream.<init>(DFSInputStream.java:270)
    at org.apache.hadoop.hdfs.DFSClient.open(DFSClient.java:1115)
    at org.apache.hadoop.hdfs.DistributedFileSystem$4.doCall(DistributedFileSystem.java:324)
    at org.apache.hadoop.hdfs.DistributedFileSystem$4.doCall(DistributedFileSystem.java:321)
    at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
    at org.apache.hadoop.hdfs.DistributedFileSystem.open(DistributedFileSystem.java:321)
Caused by: org.apache.hadoop.ipc.RemoteException(java.io.FileNotFoundException): File does not exist: /user/root/tdm/train/checkpoint/checkpoints
    at org.apache.hadoop.hdfs.server.namenode.INodeFile.valueOf(INodeFile.java:71)
    at org.apache.hadoop.hdfs.server.namenode.INodeFile.valueOf(INodeFile.java:61)
    at org.apache.hadoop.hdfs.server.namenode.FSDirStatAndListingOp.getBlockLocations(FSDirStatAndListingOp.java:152)
    at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.getBlockLocations(FSNamesystem.java:1750)
    at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.getBlockLocations(NameNodeRpcServer.java:691)
    at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.getBlockLocations(ClientNamenodeProtocolServerSideTranslatorPB.java:377)
    at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
    at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:447)
    at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:989)
    at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:845)
    at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:788)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:422)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1807)
    at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2455)

    at org.apache.hadoop.ipc.Client.getRpcResponse(Client.java:1481)
    at org.apache.hadoop.ipc.Client.call(Client.java:1427)
    at org.apache.hadoop.ipc.Client.call(Client.java:1337)
    at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:227)
    at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:116)
    at com.sun.proxy.$Proxy10.getBlockLocations(Unknown Source)
    at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.getBlockLocations(ClientNamenodeProtocolTranslatorPB.java:255)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:398)
    at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeMethod(RetryInvocationHandler.java:163)
    at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invoke(RetryInvocationHandler.java:155)
    at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeOnce(RetryInvocationHandler.java:95)
    at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:335)
    at com.sun.proxy.$Proxy11.getBlockLocations(Unknown Source)
    at org.apache.hadoop.hdfs.DFSClient.callGetBlockLocations(DFSClient.java:826)
    ... 10 more
result is None, finished success.
start put item_emb
19/08/13 12:15:03 WARN hdfs.DFSClient: zero
GetLatestCheckpoint [hdfs://{ip}:{host}/user/root/tdm/train/checkpoint//ckpt-...................2]
19/08/13 12:15:03 WARN hdfs.DFSClient: zero
Start convert [item_emb], part[0]
19/08/13 12:15:03 WARN hdfs.DFSClient: zero
finish convert [item_emb], part[0], id_cnt[0], duration[45ms]
finish convert ckpt[hdfs://{ip}:{host}/user/root/tdm/train/checkpoint/], duration[124ms]
rm -f data/item_emb
hadoop fs -get hdfs:///user/root/tdm/embed/item_emb data/item_emb
sed -i 's/..//' data/item_emb
hadoop fs -put -f data/item_emb hdfs:///user/root/tdm/embed
finish put item_emb
Exception AttributeError: "'NoneType' object has no attribute 'NNSymbolFree'" in <object repr() failed> ignored
Exception AttributeError: "'NoneType' object has no attribute 'NNSymbolFree'" in <object repr() failed> ignored
Exception AttributeError: "'NoneType' object has no attribute 'NNSymbolFree'" in <object repr() failed> ignored
Exception AttributeError: "'NoneType' object has no attribute 'NNSymbolFree'" in <object repr() failed> ignored
Exception AttributeError: "'NoneType' object has no attribute 'NNSymbolFree'" in <object repr() failed> ignored
Exception AttributeError: "'NoneType' object has no attribute 'NNSymbolFree'" in <object repr() failed> ignored
```

**Comment by lovickie** [comment]
```
>>>>>>>>>>>> 0 >>>>>>>>>>>
save checkpoint at global_step[2], ckpt version[ckpt-...................2]
```

It means you didn't read sample files correctly.

**Comment by better629** [comment]
> ```
> >>>>>>>>>>>> 0 >>>>>>>>>>>
> save checkpoint at global_step[2], ckpt version[ckpt-...................2]
> ```
> 
> It means you didn't read sample files correctly.

If the example is fine in your environment, why it didn't read sample file correctly???

The above questions are:
1、`registry.cn-hangzhou.aliyuncs.com/xdl/xdl:ubuntu-gpu-mxnet1.3` docker can't detect host machine cuda device.
2、cpu version is also not working.

**Comment by lovickie** [comment]
Check problem 2 first? It seemed to be the sample hdfs addr problem, for example there are no files in it, all file names are not right...

**Comment by better629** [comment]
@lovickie 
`java.io.FileNotFoundException: File does not exist: /user/root/tdm/train/checkpoint/checkpoints
 `
The `/user/root/tdm/train/checkpoint/`should be empty with your instruction, then the `checkpoings` is a file or a folder? If it's a file, it should have the `write` permission to hdfs.

**Comment by better629** [comment]
@lovickie

**Comment by better629** [comment]
@lovickie 
some suggestions of the instruction in `https://github.com/alibaba/x-deeplearning/wiki/深度树匹配模型(TDM)`

The project use the hdfs to store data in two different ways
1、hadoop binary command
2、third_party hdfs
so the usage of `**_HDFS_PATH`  is different. But it's unlear in instruction part.

Because we have set the hadoop environment, so the `**__HDFS_PATH` format to use can be the format like 
```
export CSV_HDFSPATH="hdfs:///user/root/tdm"
export UPLOAD_HDFSPATH="hdfs:///user/root/tdm/tree_data"
export EMB_HDFSPATH="hdfs:///user/root/tdm/embed"
export INIT_CKPT_HDFS_PATH="hdfs:///user/root/tdm/tdm_mock_init"
export TRAIN_CKPT_HDFS_PATH="hdfs:///user/root/tdm/train"
export CLUSTER_CKPT_HDFS_PATH="hdfs:///user/root/tdm/cluster"
```

But, in config.* or tdm.json files. The `**_HDFS_PATH`s like `upload_url / model_url in data/tdm.json` is used by third-party hdfs. Although the code segment in `train.py` has been set to `data_io = xdl.DataIO(namenode="hdfs://{hdfs_namenode_ip}:{hdfs_namenode_port}, ****)`. If you don't set `"upload_url": "hdfs://{hdfs_namenode_ip}:{hdfs_namenode_port}/user/root/tdm/tree_data"`, it can't recognize the right path.
So, ** please add  {hdfs_namenode_ip}:{hdfs_namenode_port} to hdfs path like above in your config files and tdm.json **

**Comment by better629** [comment]
@lovickie 
I have set `EMB_HDFSPATH `, but why I can't get the output of TDM model data `dense.txt、graph.txt、sparse.txt`.

**Comment by lovickie** [comment]
> @lovickie
> `java.io.FileNotFoundException: File does not exist: /user/root/tdm/train/checkpoint/checkpoints `
> The `/user/root/tdm/train/checkpoint/`should be empty with your instruction, then the `checkpoings` is a file or a folder? If it's a file, it should have the `write` permission to hdfs.

When xdl trains a new mission, checkpoint dir is empty, so checkpoints doesn't exist, it is not an error.

**Comment by lovickie** [comment]
> @lovickie
> some suggestions of the instruction in `https://github.com/alibaba/x-deeplearning/wiki/深度树匹配模型(TDM)`
> 
> The project use the hdfs to store data in two different ways
> 1、hadoop binary command
> 2、third_party hdfs
> so the usage of `**_HDFSPATH or **_HDFS_PATH` is different. But it's unlear in instruction part.
> 
> Because we have set the hadoop environment, so the `**__HDFSPATH or **_HDFS_PATH` format to use can be the format like
> 
> ```
> export CSV_HDFSPATH="hdfs:///user/root/tdm"
> export UPLOAD_HDFSPATH="hdfs:///user/root/tdm/tree_data"
> export EMB_HDFSPATH="hdfs:///user/root/tdm/embed"
> export INIT_CKPT_HDFS_PATH="hdfs:///user/root/tdm/tdm_mock_init"
> export TRAIN_CKPT_HDFS_PATH="hdfs:///user/root/tdm/train"
> export CLUSTER_CKPT_HDFS_PATH="hdfs:///user/root/tdm/cluster"
> ```
> 
> Particular, the format should be `hdfs:///user` not `hdfs://user`
> 
> But, in config.* or tdm.json files. The `**_HDFSPATH`s like `upload_url / model_url in data/tdm.json` is used by third-party hdfs. Although the code segment in `train.py` has been set to `data_io = xdl.DataIO(namenode="hdfs://{hdfs_namenode_ip}:{hdfs_namenode_port}, ****)`. If you don't set `"upload_url": "hdfs://{hdfs_namenode_ip}:{hdfs_namenode_port}/user/root/tdm/tree_data"`, it can't recognize the right path.
> 
> So, **please add {hdfs_namenode_ip}:{hdfs_namenode_port} to hdfs path like above in your config files and tdm.json**

Yes. Namenode should be written always.

**Comment by lovickie** [comment]
> @lovickie
> some suggestions of the instruction in `https://github.com/alibaba/x-deeplearning/wiki/深度树匹配模型(TDM)`
> 
> The project use the hdfs to store data in two different ways
> 1、hadoop binary command
> 2、third_party hdfs
> so the usage of `**_HDFSPATH or **_HDFS_PATH` is different. But it's unlear in instruction part.
> 
> Because we have set the hadoop environment, so the `**__HDFSPATH or **_HDFS_PATH` format to use can be the format like
> 
> ```
> export CSV_HDFSPATH="hdfs:///user/root/tdm"
> export UPLOAD_HDFSPATH="hdfs:///user/root/tdm/tree_data"
> export EMB_HDFSPATH="hdfs:///user/root/tdm/embed"
> export INIT_CKPT_HDFS_PATH="hdfs:///user/root/tdm/tdm_mock_init"
> export TRAIN_CKPT_HDFS_PATH="hdfs:///user/root/tdm/train"
> export CLUSTER_CKPT_HDFS_PATH="hdfs:///user/root/tdm/cluster"
> ```
> 
> Particular, the format should be `hdfs:///user` not `hdfs://user`
> 
> But, in config.* or tdm.json files. The `**_HDFSPATH`s like `upload_url / model_url in data/tdm.json` is used by third-party hdfs. Although the code segment in `train.py` has been set to `data_io = xdl.DataIO(namenode="hdfs://{hdfs_namenode_ip}:{hdfs_namenode_port}, ****)`. If you don't set `"upload_url": "hdfs://{hdfs_namenode_ip}:{hdfs_namenode_port}/user/root/tdm/tree_data"`, it can't recognize the right path.
> 
> So, **please add {hdfs_namenode_ip}:{hdfs_namenode_port} to hdfs path like above in your config files and tdm.json**

Now can you train the tdm example @cpu ?

**Comment by better629** [comment]
> Now can you train the tdm example @cpu ?
@lovickie 
It does't go well with the cpu-docker-image about att_ubuntu example.

error:
```
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
if self.use_batch_norm:
if self.active_op == 'prelu':
if mx.symbol.FullyConnected
[WARNING]: no gradient found for var:bn_1_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_1_d_moving_var under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_2_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_2_d_moving_var under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_3_d_moving_mean under scope: , maybe not used?
[WARNING]: no gradient found for var:bn_3_d_moving_var under scope: , maybe not used?
>>>>>>>>>>>> 0 >>>>>>>>>>>
19/08/20 15:45:40 WARN hdfs.DFSClient: zero
Traceback (most recent call last):
  File "train.py", line 331, in <module>
    train(is_training=True)
  File "train.py", line 281, in train
    result = sess.run([train_op, xdl.get_collection(xdl.UPDATE_OPS)])
  File "/usr/local/lib/python2.7/dist-packages/xdl/python/training/train_session.py", line 367, in run
    return self._session.run(v, run_option, run_statistic)
  File "/usr/local/lib/python2.7/dist-packages/xdl/python/training/train_session.py", line 343, in run
    return self._session.run(v, run_option, run_statistic)
  File "/usr/local/lib/python2.7/dist-packages/xdl/python/framework/session.py", line 55, in run
    results = execute(run_item, run_option, run_statistic)
  File "/usr/local/lib/python2.7/dist-packages/xdl/python/lib/graph.py", line 278, in execute
    return current_graph().execute(outputs, run_option, run_statistic)
  File "/usr/local/lib/python2.7/dist-packages/xdl/python/lib/graph.py", line 210, in execute
    result = pybind.execute(self._graph_def, xdl_output_spec, run_option)
RuntimeError: [15:45:40] /usr/local/lib/python2.7/dist-packages/mxnet-1.3.0-py2.7.egg/mxnet/cpp-package/include/mxnet-cpp/symbol.hpp:
```

xdl version=1.2.
Above problem has occured ini mxnet=1.3.1 and fixed with mxnet=1.3.0. I have no idea why happens here.
